### PR TITLE
Time Procedures and Drug ETLs

### PR DESCRIPTION
**Pull request by labkey-bpatel**
#### Rationale
Implement incremental ETL queries for Timed Procedures (Vitals & Drug 'Sedation' event) and Drug ('Behavioral Treatment, Dose Administration, Fluid Administration, Medication Administration, Oral Dose Administration - Alert, Sedation, Vaccinations' events)

#### Related Pull Requests
* https://github.com/LabKey/dataintegration/pull/128

#### Changes
* Update Vitals query to include Time Procedure events
* Add Drug dataset, etl setup, report, etl queries
* Add userGroupName constant to etl config in order to create users in LabKey

### CODE REVIEW

File: nirc_ehr/resources/referenceStudy/datasets/datasets_manifest.xml [hunk 1]
```diff
@@ -15,6 +15,7 @@
         <dataset name="demographics" id="1012" category="Colony Management" demographicData="true" type="Standard"/>
         <dataset name="departure" id="1013" category="Colony Management" type="Standard"/>
         <dataset name="exemptions" id="1016" category="Colony Management" type="Standard"/>
+        <dataset name="drug" id="1016" category="Clinical" type="Standard"/>
```
Comment (ankurjuneja) [inline nirc_ehr/resources/referenceStudy/datasets/datasets_manifest.xml]:
```suggestion
        <dataset name="drug" id="1017" category="Clinical" type="Standard"/>
```

File: nirc_ehr/resources/queries/study/drug/.qview.xml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+<customView xmlns="http://labkey.org/data/xml/queryCustomView">
+    <columns>
```
Comment (ankurjuneja) [inline nirc_ehr/resources/queries/study/drug/.qview.xml:2]: suggest to apply a descending sort on date.

**Review by ankurjuneja — APPROVED** [review]

File: nirc_ehr/resources/referenceStudy/datasets/datasets_manifest.xml (hunk 1 above)
Comment (labkey-jeckels) [reply nirc_ehr/resources/referenceStudy/datasets/datasets_manifest.xml]: @ankurjuneja have you tried removing the ID attribute completely from these XML files? I don't think we actually care about the IDs, and it's not required per the XSD.

Comment (ankurjuneja) [reply nirc_ehr/resources/referenceStudy/datasets/datasets_manifest.xml]: No, I didn't. Will try that and remove all of them, Thanks.

File: nirc_ehr/resources/queries/dbo/q_drug.sql [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+--     3	Vaccinations
+--     25	Dose Administration
+--     34	Medication Administration
+--     39	Behavioral Treatment
+--     51	Oral Dose Administration - Alert
+--     54	Fluid Administration
+--     65	Sedation
+AND anmEvt.CREATED_DATETIME < now() -- there are rows in ANIMAL_EVENT table with future dates
+
+GROUP BY anmEvt.ANIMAL_EVENT_ID,
```
Comment (labkey-jeckels) [inline nirc_ehr/resources/queries/dbo/q_drug.sql:31]: Why is this GROUP BY necessary? I don't see any SUM/MAX/MIN or other aggregates above? Is there duplicated data?

File: nirc_ehr/resources/referenceStudy/datasets/datasets_manifest.xml (hunk 1 above)
Comment (labkey-bpatel) [reply nirc_ehr/resources/referenceStudy/datasets/datasets_manifest.xml]: I can do that as part of the code review in this fb

File: nirc_ehr/resources/queries/study/drug/.qview.xml (hunk 2 above)
Comment (labkey-bpatel) [reply nirc_ehr/resources/queries/study/drug/.qview.xml:2]: good point, done!

File: nirc_ehr/resources/queries/dbo/q_drug.sql (hunk 3 above)
Comment (labkey-bpatel) [reply nirc_ehr/resources/queries/dbo/q_drug.sql:31]: There are duplicate rows which are due to the join to event_event_groups table, where multiple (drug event) groups have the same event associated with it.
Example:
Event Group Id:         Event Id:
25                              1608
34                              1608
51                              1608

File: nirc_ehr/resources/referenceStudy/datasets/datasets_manifest.xml (hunk 1 above)
Comment (labkey-bpatel) [reply nirc_ehr/resources/referenceStudy/datasets/datasets_manifest.xml]: Removing id attribute from dataset_manifest.xml file throws this exception:

ERROR ExceptionUtil            2021-10-04T14:52:45,553     http-nio-8088-exec-2 : Exception detected and logged to mothership with error code: NUVN8I
org.labkey.api.admin.ImportException: Invalid DatasetsDocument 
	at org.labkey.study.importer.DatasetDefinitionImporter.getDatasetsManifest(DatasetDefinitionImporter.java:241) ~[study-21.7-SNAPSHOT.jar:?]
	at org.labkey.study.importer.DatasetCohortAssigner.isValidForImportArchive(DatasetCohortAssigner.java:106) ~[study-21.7-SNAPSHOT.jar:?]
	at org.labkey.study.importer.DatasetCohortAssigner.isValidForImportArchive(DatasetCohortAssigner.java:41) ~[study-21.7-SNAPSHOT.jar:?]
	at org.labkey.study.importer.StudyImporterFactory$StudyFolderImporter.getChildrenDataTypes(StudyImporterFactory.java:175) ~[study-21.7-SNAPSHOT.jar:?]
	at org.labkey.core.CoreController$GetRegisteredFolderImportersAction.execute(CoreController.java:1870) ~[core-21.7-SNAPSHOT.jar:?]
	at org.labkey.core.CoreController$GetRegisteredFolderImportersAction.execute(CoreController.java:1842) ~[core-21.7-SNAPSHOT.jar:?]
	at org.labkey.api.action.BaseApiAction.handlePost(BaseApiAction.java:214) [api-21.7-SNAPSHOT.jar:?]
	at org.labkey.api.action.BaseApiAction.handleRequest(BaseApiAction.java:129) [api-21.7-SNAPSHOT.jar:?]
	at org.labkey.api.action.BaseViewAction.handleRequest(BaseViewAction.java:176) [api-21.7-SNAPSHOT.jar:?]
	at org.labkey.api.action.SpringActionController.handleRequest(SpringActionController.java:511) [api-21.7-SNAPSHOT.jar:?]
	at org.labkey.api.module.DefaultModule.dispatch(DefaultModule.java:1117) [api-21.7-SNAPSHOT.jar:?]
	at org.labkey.api.view.ViewServlet._service(ViewServlet.java:218) [api-21.7-SNAPSHOT.jar:?]
	at org.labkey.api.view.ViewServlet.service(ViewServlet.java:142) [api-21.7-SNAPSHOT.jar:?]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:733) [servlet-api.jar:4.0.FR]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:227) [catalina.jar:9.0.44]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) [catalina.jar:9.0.44]
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53) [tomcat-websocket.jar:9.0.44]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) [catalina.jar:9.0.44]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) [catalina.jar:9.0.44]
	at org.labkey.api.data.TransactionFilter.doFilter(TransactionFilter.java:40) [internal-21.7-SNAPSHOT.jar:?]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) [catalina.jar:9.0.44]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) [catalina.jar:9.0.44]
	at org.labkey.api.module.ModuleLoader.doFilter(ModuleLoader.java:1307) [api-21.7-SNAPSHOT.jar:?]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) [catalina.jar:9.0.44]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) [catalina.jar:9.0.44]
	at org.labkey.api.security.AuthFilter.doFilter(AuthFilter.java:218) [internal-21.7-SNAPSHOT.jar:?]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) [catalina.jar:9.0.44]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) [catalina.jar:9.0.44]
	at org.labkey.core.filters.SetCharacterEncodingFilter.doFilter(SetCharacterEncodingFilter.java:120) [core-21.7-SNAPSHOT.jar:?]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) [catalina.jar:9.0.44]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) [catalina.jar:9.0.44]
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:202) [catalina.jar:9.0.44]
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97) [catalina.jar:9.0.44]
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:542) [catalina.jar:9.0.44]
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:143) [catalina.jar:9.0.44]
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92) [catalina.jar:9.0.44]
	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:687) [catalina.jar:9.0.44]
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78) [catalina.jar:9.0.44]
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:357) [catalina.jar:9.0.44]
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:374) [tomcat-coyote.jar:9.0.44]
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65) [tomcat-coyote.jar:9.0.44]
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:893) [tomcat-coyote.jar:9.0.44]
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1707) [tomcat-coyote.jar:9.0.44]
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49) [tomcat-coyote.jar:9.0.44]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) [?:?]
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) [tomcat-util.jar:9.0.44]
	at java.lang.Thread.run(Thread.java:832) [?:?]
Caused by: org.labkey.api.util.XmlValidationException: Document does not conform to its XML schema, D=datasets@http://labkey.org/study/xml
	at org.labkey.api.util.XmlBeansUtil.validateXmlDocument(XmlBeansUtil.java:82) ~[api-21.7-SNAPSHOT.jar:?]
	at org.labkey.api.util.XmlBeansUtil.validateXmlDocument(XmlBeansUtil.java:71) ~[api-21.7-SNAPSHOT.jar:?]
	at org.labkey.study.importer.DatasetDefinitionImporter.getDatasetsManifest(DatasetDefinitionImporter.java:234) ~[study-21.7-SNAPSHOT.jar:?]
	... 47 more

Comment (labkey-jeckels) [reply nirc_ehr/resources/referenceStudy/datasets/datasets_manifest.xml]: Bummer. IntelliJ ctrl-clicked me into the wrong XSD. It's required in datasets.xsd, hence this error. I'd like to make it optional in the future, but for now we'll have to retain the IDs (and make them unique)

Comment (labkey-bpatel) [reply nirc_ehr/resources/referenceStudy/datasets/datasets_manifest.xml]: sounds good! Thanks

**Review by labkey-jeckels — APPROVED** [review]

**Pull request merged**